### ISSUE TITLE
QGIS crashes when layers added / removed / renamed from WMS provider

### ISSUE THREAD

**Issue by beedawg1985** [qgis/QGIS#39476]
**Describe the bug**

QGIS crashes when layers added / removed / renamed from WMS provider

**How to Reproduce**

- Add a new Geoserver WMS connection to QGIS from Browser window
- Alter layers within Geoserver (i.e remove a layer)
- Attempt to refresh QGIS connection in Browser window
- Crash

(Same workflow also causes crash when handling layers from a Geopackage)

**QGIS and OS versions**

QGIS version | 3.10.10-A Coruña | QGIS code revision | 1869829378
-- | -- | -- | --
Compiled against Qt | 5.12.3 | Running against Qt | 5.12.3
Compiled against GDAL/OGR | 2.4.1 | Running against GDAL/OGR | 2.4.1
Compiled against GEOS | 3.7.2-CAPI-1.11.2 | Running against GEOS | 3.7.2-CAPI-1.11.2 b55d2125
Compiled against SQLite | 3.28.0 | Running against SQLite | 3.28.0
PostgreSQL Client Version | 11.3 | SpatiaLite Version | 4.3.0a
QWT Version | 6.1.4 | QScintilla2 Version | 2.11.1
Compiled against PROJ | 5.2.0 | Running against PROJ | Rel. 5.2.0, September 15th, 2018
OS Version | macOS 10.15
Active python plugins | DigitizingTools; QuickOSM; profiletool; joinmultiplelines; SLD4raster; HCMGIS; TerrainShading; OSMDownloader; openlayers_plugin; batch_hillshader-master; mmqgis; getools; osgb-plugin; send2google_earth; processing; db_manager; MetaSearch

**Comment by elpaso** [comment]
don't you have a stack trace?

**Comment by roya0045** [comment]
Is this valid on the latest version?

**Comment by beedawg1985** [comment]
> don't you have a stack trace?

For some reason the stack trace didn't pop up following the last crash. The bug should be easy enough to replicate; it also affects geopackage databases.

**Comment by roya0045** [comment]
Did you try with a clean profile? Mac specific?

**Comment by beedawg1985** [comment]
I will try with a clean profile — guest user is OK I'm guessing? I haven't tested on other OS as yet, so Mac specific for now.

**Comment by gioman** [comment]
I was able to replicate:

```
qgis
Warning 1: libXML: 0: Element 'SrcSql': This element is not expected.
DEBUG:matplotlib:CACHEDIR=/home/giovanni/.cache/matplotlib
DEBUG:matplotlib.font_manager:Using fontManager instance from /home/giovanni/.cache/matplotlib/fontlist-v310.json
DEBUG:matplotlib.pyplot:Loaded backend qt5agg version unknown.
DEBUG:matplotlib.pyplot:Loaded backend Qt5Agg version unknown.
Warning 4: Failed to open /usr/share/qgis/resources/data/world_map.gpkg: Permission denied.
QGIS died on signal 11Could not attach to process.  If your uid matches the uid of the target
process, check the setting of /proc/sys/kernel/yama/ptrace_scope, or try
again as the root user.  For more details, see /etc/sysctl.d/10-ptrace.conf
ptrace: Operation not permitted.
No thread selected
No stack.
gdb returned 256
Aborted (core dumped)
```


but interestingly:

1) I was not able to replicate when removing  a layer from the WMS service
2) I was able to replicate after *adding** a new layer to the WMS service
3) clicking the "refresh" button on the browser do not seems to trigger any refresh at all (after having changed the remote service)
4) I was able to get the crash only after having added a new layer, open the "add WMS layer" dialog and clicked on "connect", and only then clicked on "refresh" in the browser.

QGIS 3.10.10 on Ubuntu 20.04.

**Comment by Samweli** [comment]
Hi @gioman, @beedawg1985, I provided a fix for a similar issue here https://github.com/qgis/QGIS/pull/33651, it available in QGIS  3.12.0

**Comment by gioman** [comment]
> Hi @gioman, @beedawg1985, I provided a fix for a similar issue

@Samweli the patch was merge in January in master, this means that it should be in any release like 3.10 and above. This means that this issue is a different one(?).

**Comment by Samweli** [comment]
> @Samweli the patch was merge in January in master, this means that it should be in any release like 3.10 and above. This means that this issue is a different one(?).

I think there was no backport for it, so it is only available starting on 3.12.0 onwards.

**Comment by gioman** [comment]
> I think there was no backport for it, so it is only available starting on 3.12.0 onwards.

ok so this must be retested on 3.16, will do it.

**Comment by gioman** [comment]
> but interestingly:
> 
>     1. I was not able to replicate when removing  a layer from the WMS service
> 
>     2. I was able to replicate after **adding** a new layer to the WMS service
> 
>     3. clicking the "refresh" button on the browser do not seems to trigger any refresh at all (after having changed the remote service)
> 
>     4. I was able to get the crash only after having added a new layer, open the "add WMS layer" dialog and clicked on "connect", and only then clicked on "refresh" in the browser.

@Samweli no crash anymore for me in the above scenario on 3.16 but it still does not refresh the list of layers in the browser, unless first re-connecting to the service using "add WMS layer" dialog.

**Comment by gioman** [comment]
Closing?

**Comment by Samweli** [comment]
> @Samweli no crash anymore for me in the above scenario on 3.16 but it still does not refresh the list of layers in the browser, unless first re-connecting to the service using "add WMS layer" dialog

ok @gioman, may be we could open another issue for that so I can take a deep look at it.

**Comment by gioman** [comment]
> ok @gioman, may be we could open another issue for that so I can take a deep look at it.

@Samweli done here https://github.com/qgis/QGIS/issues/40350 thanks!
